[NNVM][TEST] Test against numerical grad

sgrechanik-h opened the pull request:
This PR moves the helper function checking nnvm operations against numpy reference implementations from `test_top_level#` to `nnvm.testing` and adds the ability to compare symbolic gradients to numerically computed gradients. Also this PR fixes a wrong gradient implementation for the division operation (it was found using this approach).

Currently only tests that used the old helper function are rewritten in such a way that they now use the new function. There are still many tests which didn't use the helper function, they should also be rewritten so that they use this function, if possible, but this is left for future work.

sgrechanik-h commented:
@kazum @kevinthesun @srkreddy1238 could you please review this PR?

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+    shape = shape.copy() if shape else {}
+
+    grad_input_vars_real = []
+    for x in grad_input_vars:
```
This logic looks complicated. How about just passing a list of symbol as grad_input_vars? Shape information can be defined in shape(maybe input_shapes of dict from str to tuple?) argument so that we don't have multiple shapes.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+    dtypes = graph.json_attr('dtype')
+
+    if shapes is None or dtypes is None:
+        graph = graph.apply('InferShape').apply('InferType')
```
Here we assume set_shape_inputs and set_dtype_inputs have been called. It might be better to directly pass input_shapes and input_types into this function to make it more general.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+
+    forward_graph = nnvm.graph.create(symbol)
+
+    if dtype is not None:
```
We need to set_dtype_inputs. just check dtype is not None.

sgrechanik-h replied:
Sorry, I think I didn't understand this one.

kevinthesun replied:
If dtype is None, graph_to_function will raise exception? Same thing will happen for shape.
How about setting dtype as optional but not None as default value, such as dtype="float32"? For shape argument we should force user to pass in value. Then the code logic can be simpler.

sgrechanik-h replied:
Setting dtype='float32' as default would simplify a lot of tests a little bit. However, if some variables have their dtypes specified by attributes then either these attributes should be preferred or an error should be raised, so I'm not sure.

kevinthesun replied:
OK. Maybe we can put the shape/dtype check right in this function, before infer shape/dtype, and check either input variable shape/dtype attributes are set or shape/dtype arguments are passed.

sgrechanik-h replied:
Since shapes and dtypes of some inputs may be inferred from shapes and dtypes of other inputs, the proper solution would be to run inference passes, and only then check if some inputs are left without types. Then these inputs can be assigned the default type.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+    if dtype is not None:
+        nnvm.compiler.graph_attr.set_dtype_inputs(forward_graph, dtype)
+
+    if shape is not None:
```
Same here.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+    forward_graph = forward_graph.apply('InferShape').apply('InferType')
+    shapes = forward_graph.json_attr('shape')
+    dtypes = forward_graph.json_attr('dtype')
+    out_shape = shapes[forward_graph.index.output_entries[0][0]]
```
Multiple outputs?

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+    # create the graph twice
+    if np_backward is not None or numerical_grads:
+        try:
+            head_grads_symbol = nnvm.symbol.Variable("head_grads", shape=out_shape, dtype=out_dtype)
```
This function just supports computing grads of single output. Maybe we should support multiple outputs as well.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+                                                grad_ys=head_grads_symbol)
+            # Sometimes grads do not depend on head_grads, so head_grads does not appear
+            # in the variable list; adding it manually prevents this, making things a bit easier
+            backward_graph = \
```
use graph.apply("Gradient") to create backward graph, similar to legacy helper function.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+            if shape is not None:
+                nnvm.compiler.graph_attr.set_shape_inputs(backward_graph, shape)
+
+            backward_graph = backward_graph.apply('InferShape').apply('InferType')
```
With graph.apply("Gradient"), we don't need to infershape/dtype again.

kevinthesun replied:
https://github.com/dmlc/tvm/blob/master/nnvm/tests/python/compiler/test_top_level1.py#L32-L37
We just need to update the shape head_grads for input shape dictionary when compiling.

sgrechanik-h replied:
Well, I perform shape and type inference here for two reasons: it makes it easier to query shape and dtype information during random input generation, and it is nice to have a graph printed out with inferred shapes and dtypes when the function fails. Of course, it is possible to avoid the second shape and dtype inference by manually adding head_grads to the list of variables we want to generate data for, but I see no benefits.

kevinthesun replied:
OK. This looks good to me.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+        low = -1.0
+        high = 1.0
+        if in_range is not None:
+            if isinstance(in_range, dict):
```
If it is just for random ndarray generation, it might be better to just set in_range as (low, high) directly? Will it make much difference for different in_range for inputs? This can make the logic much simpler.

Or we can simply define it as dict from str to num array.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+            # nnvm_res contains the output and gradients (if they are needed)
+            nnvm_res = main_function(**np_inputs)
+
+            if np_forward is not None:
```
If np_forward is None, forward part comparison will be skipped? I think we should enforce np_forward to be set.

sgrechanik-h replied:
Sometimes we might want check gradients only (or even only numerical gradients). But I think it is good idea to require passing `numerical_grads=True` explicitly if no reference function is specified.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+                # Since the result may be non-scalar, we have to put another operation on the top,
+                # so we just multiple by the randomly generated head_grads.
+                # This way we can reuse the gradient values which has been already computed.
+                def function(**kwargs):
```
Change to a more meaningful name.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+    ----------
+    function
+        A function that takes inputs as keyword arguments (like `function(**input_values)`) and
+        returns a scalar result. Should accept and return numpy arrays.
```
Description is not quite clear here. Scalar or ndarray?

srkreddy1238 commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+        None (default) means that all input variables will be used. May be a
+        list of pairs `(var_name, shape)`.
+
+    np_forward : Callable[..., List[numpy.ndarray]], optional
```
Suggest to give different naming for np_forward : Need not be numpy always, could be an implementation in any of the frontends we have.

Also receive and pass a flexible parameters 'attr' (dictionary object with default value None) for this function which can carry the implementation specific data like session info ...etc.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+        Relative tolerance.
+    """
+
+    if function_value is None:
```
We can just compute this value instead of setting it as an argument?

srkreddy1238 commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+                logging.debug(debug_stage)
+
+                numpy_res = np_forward(**np_inputs_without_head_grads)
+                np.testing.assert_allclose(nnvm_res[0], numpy_res, atol=atol, rtol=rtol)
```
In the context of above comment on np_forward, 
Pls check if the result from np_forward is a list and is yes, then compare all arrays in list.

srkreddy1238 commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+
+    dump_graph : bool
+        Dump the graph even on success.
+    """
```
Not sure an option to custom comparison function argument make any sense here. 
Just raising as it came across my thought.

kevinthesun commented on nnvm/tests/python/compiler/test_top_level1.py:
```diff
@@ -64,8 +22,8 @@ def backward(head_grads, x):
-    inputs = [('x', dshape, x)]
-    helper(y, inputs, dtype, forward, backward)
+    inputs = [(x, dshape)]
+    check_function(y, inputs, forward, backward, dtype=dtype)
```
Here we are setting grad_input_vars=inputs, but inputs are actually input_shapes. It would be more clear if we can refactor the interface of check_function.

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+def check_numerical_grads(function, grad_input_vars, input_values, grad_values, function_value=None,
+                          delta=1e-3, max_error=1e+3, max_discarded_frac=0.1,
+                          atol=1e-2, rtol=1e-2):
+    """A helper function that checks that numerical gradients of a function are equal to
```
Since this is new and will affect all backward operators, can you give more explanation on it?

If developers add new backward ops but finds the checking failure, how can they quickly debug the issue? Since this is not simply comparing with np computation result, maybe we can have a tutorial for it. This would also be good to the stability of CI while we are introducing more and more backward ops.

kazum commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+        if grad.shape != input_values[x_name].shape:
+            raise AssertionError(
+                "Gradient wrt '{}' has unexpected shape {}, expected {} "
+                .format(x.attr('name'), grad.shape, input_values[x_name].shape))
```
x is not defined here.  x.attr('name') should be x_name.

kazum commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+
+
+def check_numerical_grads(function, grad_input_vars, input_values, grad_values, function_value=None,
+                          delta=1e-3, max_error=1e+3, max_discarded_frac=0.1,
```
The default value of delta is different from the check_function's one.  Should be 1e-2?

sgrechanik-h replied:
Oh, it should be 1e-3 everywhere, thank you!

kazum commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+        return (function(**modified_values) - function_value)/a_delta
+
+    for x_name in grad_input_vars:
+        grad = grad_values[x_name]
```
We can replace these two lines with `for x_name, grad in grad_values.items():`.  Then `grad_input_vars` is no longer necessary.

kazum commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,451 @@
+        raise ValueError("numerical_grads must be a bool or 'if_possible', not {}"
+                         .format(numerical_grads))
+
+    input_vars = [x for x in symbol.list_input_variables()]
```
Can we simply replace this with `input_vars = symbol.list_input_variables()`?

sgrechanik-h commented:
- Renamed `np_forward` to `forward`, same for `backward`, and now it is possible to pass additional
  parameters to these functions via `additional_params`.
- Added support for multiple inputs.
- Now additional parameters to the numgrad checking function are passed as a single dictionary.

- I'm not sure yet about merging the behaviour of `numerical_grads=True` and
  `numerical_grads='if_possible'`, so I left it as it is for now.
- Require passing `numerical_grads=True` explicitly if no reference function is specified.
- Got rid of explicit passing of input variables in most of the tests. Also shapes cannot be
  passed with `grad_input_vars` now.

Also added some tests for `check_function` itself.

The main thing that is left to do is something like a tutorial on what to do if numerical gradient
testing fails. I'll return to this task in a few days.

sgrechanik-h commented:
Simplified numerical grad checking, improved docs. Wasn't sure where to put the text on what to do when numgrad check fails, so I put it under Python API/NNVM API/nnvm.testing

@kazum @kevinthesun @srkreddy1238 could you please review this again?

kevinthesun commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,529 @@
+
+    shape : Dict[nnvm.Symbol or str, Tuple[int]] or Tuple[int], optional
+        A dict mapping input variable names to shapes, or just a single shape.
+        By default shapes will be inferred automatically.
```
How can we infer input shapes? I think this argument shouldn't be optional.

sgrechanik-h replied:
They are inferred from variables' attributes:

    x = sym.Variable("x", shape=(1, 2))

kevinthesun replied:
This interface doesn't have any constraint to input symbol. So I can write some codes like:
data = sym.Variable("data")
net = sym.conv2d(data, ...)
check_function(net, shape=None, ...)

It will raise exception later, since no shape information is provided? Currently this kind of input is marked as valid.

sgrechanik-h replied:
Oh, the error message is not nice in this case. I think making it nicer and clarifying the documentation should be enough.

kazum commented on docs/api/python/nnvm/testing_new_ops.rst:
```diff
@@ -0,0 +1,135 @@
+Testing new operations
+----------------------
+
+When adding new operations, it it a good idea to test them. Testing
```
“it it” => “it is”

kazum commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,546 @@
+            all_shapes = graph.json_attr('shape')
+            all_dtypes = graph.json_attr('dtype')
+
+        all_dtypes = [None if t == -1 else TCODE_TO_DTYPE[t] for t in all_dtypes]
```
TCODE_TO_DTYPE[-1] is defined as None in graph_attr.py.  The if-else expression is not necessary.

kazum commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,546 @@
+    if None in dtype.values():
+        raise ValueError("Input variables with no type: {}".format(dtype))
+
+    if any([not s for s in shape.values()]):
```
`not all(shape.values())` looks simpler.

kazum commented on nnvm/python/nnvm/testing/check_computation.py:
```diff
@@ -0,0 +1,546 @@
+        if (exclude_targets is not None and (target in exclude_targets or
+                                             str(target) in exclude_targets)) or\
+           (only_targets is not None and not (target in only_targets or
+                                              str(target) in only_targets)):
```
The expression is a bit complicated.  I’d suggest splitting them:
```
if exclude_targets is not None:
    if target in exclude_targets or str(target) in exclude_targets:
        logging.info(...)
        continue
if only_targets is not None:
    if target not in only_targets and str(target) not in only_targets:
        logging.info(...)
        continue
```

yzhliu commented:
@srkreddy1238 @kevinthesun @kazum could you help to review again according to the new changes and explicitly approve if it is good to you.
@sgrechanik-h Please rebase.

yzhliu commented:
Thanks everyone. This is now merged.
